src/Common/src/TypeSystem/Common/TypeSystemHelpers.cs

```diff
@@ -36,18 +36,24 @@ static public PointerType MakePointerType(this TypeDesc type)
 
         static public DefType GetClosestDefType(this TypeDesc type)
         {
-            if (type is DefType)
-                return (DefType)type;
-            else
-                return type.BaseType;
+            // The implementation for this should mirror GetClosestMetadataType
+            throw new NotImplementedException();
         }
 
         static public MetadataType GetClosestMetadataType(this TypeDesc type)
         {
-            if (type is MetadataType)
-                return (MetadataType)type;
-            else
-                return type.BaseType.GetClosestMetadataType();
+            if (type.IsSzArray)
+            {
+                MetadataType arrayType = type.Context.SystemModule.GetType("System", "Array`1");
```

Comment:
The proposed design would not prevent us to do the ICastable thing. The user of the type system would be in full control of providing the interfaces and the virtual methods implementing the interfaces. They're is free to provide empty lists.

The reason why I want it to be in a single place is because it lets us easily play with (optional) whole program optimizations that can omit the interfaces if we can prove they're not needed for a particular array instance. The fewer places are forced to be aware of ArrayOfT, the easier it is to do it. Design we are choosing will make it messy because the type system won't provide hooks to hide the optimization.

But I see I'm being overvoted on this. I'll implement the other proposal when I get to it.